spec.bs

```diff
@@ -13,3 +13,538 @@ Group: patcg
+          1. Set |topic|["{{BrowsingTopic/topic}}"] to |topTopicWithCallerOrigins|'s [=topic with caller origins/topic id=].
+      1. If |topic| is null, or if |topic|'s {{BrowsingTopic/topic}} is 0 (i.e. the candidate topic was cleared), then continue.
+      1. Let |randomOrTopTopicDecisionHash| = 64 bit truncation of the output of the [=HMAC algorithm=], given input parameters: whichSha=SHA256, key=user agent's [=user agent/user topics state=]'s [=user topics state/hmac key=], and message_array=the concatenation of "random-or-top-topic-decision|", |epoch|'s [=epoch/time=], and |callerContext|'s [=topics caller context/top level context domain=].
+      1. If |randomOrTopTopicDecisionHash| % 100 < 5:
+          1. Let |randomTopicIndexDecisionHash| = 64 bit truncation of the output of the [=HMAC algorithm=], given input parameters: whichSha=SHA256, key=user agent's [=user agent/user topics state=]'s [=user topics state/hmac key=], and message_array=the concatenation of "random-topic-index-decision|", |epoch|'s [=epoch/time=], and |callerContext|'s [=topics caller context/top level context domain=].
+          1. Let |randomTopicIndex| = |randomTopicIndexDecisionHash| % |epoch|'s [=epoch/taxonomy=]'s size.
+          1. Set |topic|'s {{BrowsingTopic/topic}} to |epoch|'s [=epoch/taxonomy=][|randomTopicIndex|].
+      1. Set |topic|["{{BrowsingTopic/configVersion}}"] to to |epoch|'s [=epoch/config version=].
+      1. Set |topic|["{{BrowsingTopic/modelVersion}} to"] |epoch|'s [=epoch/model version=].
+      1. Set |topic|["{{BrowsingTopic/taxonomyVersion}}"] to |epoch|'s [=epoch/taxonomy version=].
+      1. Determine the [=browsing topics types/version=] |version|, given |topic|'s {{BrowsingTopic/configVersion}}, {{BrowsingTopic/modelVersion}} and {{BrowsingTopic/taxonomyVersion}} as input.
+      1. Set |topic|["{{BrowsingTopic/version}}"] to |version|.
+      1. Add |topic| to |result|.
+    1. Sort entries in |result| given the <a href="#browsing-topics-dictionary-less-than-comparator">less-than comparator</a> for the {{BrowsingTopic}} dictionary.
+    1. Remove duplicate entries in |result|.
+    1. Return |result|.
+  </div>
+
+</section>
+
+<section>
+  <h2 id="the-javascript-api-header">The JavaScript API</h2>
+
+  The Topics API lives under the {{Document}} interface, and is only available if the document is in [=secure context=].
+
+  <pre class="idl">
+    dictionary BrowsingTopicsOptions {
+      boolean skipObservation = false;
+    };
+
+    partial interface Document {
+        Promise&lt;sequence&lt;BrowsingTopic&gt;&gt; browsingTopics(optional BrowsingTopicsOptions options = {});
+    };
+  </pre>
+
+  <div algorithm="browsingTopics(options)">
+    The <dfn for="Document" method>browsingTopics(options)</dfn> method steps are:
+    1. Let |document| be [=this=].
+    1. Let |promise| be [=a new promise=].
+    1. Run the following steps [=in parallel=]:
+      1. Let |topicsCallerContext| be a [=topics caller context=].
```

Comment:
You should construct this and fill its fields before going in parallel. Someone could change the origin or domain out from under you, by using `document.domain` on the main thread.